internal/armada/server/submit_to_log.go

```diff
@@ -158,21 +170,74 @@ func (srv *PulsarSubmitServer) SubmitJobs(ctx context.Context, req *api.JobSubmi
 
 	// Incoming gRPC requests are annotated with a unique id.
 	// Pass this id through the log by adding it to the Pulsar message properties.
-	requestId := requestid.FromContextOrMissing(ctx)
-	msg := &pulsar.ProducerMessage{
-		Payload:    payload,
-		Properties: map[string]string{requestid.MetadataKey: requestId},
-		Key:        req.JobSetId,
-	}
-	pulsarrequestid.AddToMessage(msg, requestId)
-	_, err = srv.Producer.Send(ctx, msg)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to send message")
+	if len(sequence.Events) > 0 {
+		requestId := requestid.FromContextOrMissing(ctx)
+		msg := &pulsar.ProducerMessage{
+			Payload:    payload,
+			Properties: map[string]string{requestid.MetadataKey: requestId},
+			Key:        req.JobSetId,
+		}
+		pulsarrequestid.AddToMessage(msg, requestId)
+		_, err = srv.Producer.Send(ctx, msg)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to send message")
+		}
 	}
 
 	return &api.JobSubmitResponse{JobResponseItems: responses}, nil
 }
 
+// removeDuplicateSubmissions filters out job submit request items with previously seen client ids.
+// Client ids are namespaced by queue.
+func (srv *PulsarSubmitServer) removeDuplicateSubmissions(ctx context.Context, queue string, items []*api.JobSubmitRequestItem) ([]*api.JobSubmitRequestItem, error) {
+
+	// For storage efficiency, we store hashes instead of user-provided strings.
+	// For computational efficiency, we create a lookup table to avoid computing the same hash twice.
+	// Client ids are namespaced by queue. Hence, we hash the client id together with the queue.
+	jobSubmitRequestItems := make([]*api.JobSubmitRequestItem, 0, len(items))
+	combinedHashData := make([]byte, 40)
```

Comment:
is 40 bytes enough here?

Reply:
Yes, the compiler guarantees that each hash is exactly 20 bytes.

Reply:
ALso if you're just worried about size in memory then string interning would probably be better.  If you're worried about the size on disk in the db then hashes work